common/txmgr/tracker.go

```diff
@@ -99,75 +96,85 @@ func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Start(ctx c
 
 	if err := tr.setEnabledAddresses(ctx); err != nil {
 		return fmt.Errorf("failed to set enabled addresses: %w", err)
 	}
-	tr.lggr.Info("Enabled addresses set")
-
-	if err := tr.trackAbandonedTxes(ctx); err != nil {
-		return fmt.Errorf("failed to track abandoned txes: %w", err)
-	}
-
-	tr.isStarted = true
-	if len(tr.txCache) == 0 {
-		tr.lggr.Info("no abandoned txes found, skipping runLoop")
-		return nil
-	}
+	tr.lggr.Infof("enabled addresses set for chainID %v", tr.chainID)
 
-	tr.lggr.Infof("%d abandoned txes found, starting runLoop", len(tr.txCache))
+	tr.chStop = make(chan struct{})
 	tr.wg.Add(1)
-	go tr.runLoop()
+	go tr.runLoop(tr.chStop.NewCtx())
+	tr.isStarted = true
 	return nil
 }
 
 func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Close() error {
-	tr.lock.Lock()
-	defer tr.lock.Unlock()
 	return tr.StopOnce("Tracker", func() error {
 		return tr.closeInternal()
 	})
 }
 
 func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) closeInternal() error {
+	tr.initSync.Lock()
+	defer tr.initSync.Unlock()
+
 	tr.lggr.Info("stopping tracker")
 	if !tr.isStarted {
-		return fmt.Errorf("tracker not started")
+		return fmt.Errorf("tracker is not started: %w", services.ErrAlreadyStopped)
 	}
-	tr.ctxCancel()
+
+	close(tr.chStop)
 	tr.wg.Wait()
 	tr.isStarted = false
 	return nil
 }
 
-func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) runLoop() {
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) runLoop(ctx context.Context, cancel context.CancelFunc) {
 	defer tr.wg.Done()
+	defer cancel()
+
+	if err := tr.trackAbandonedTxes(ctx); err != nil {
+		tr.lggr.Errorf("failed to track abandoned txes: %v", err)
+		return
+	}
+	if err := tr.handleTxesByState(ctx, 0); err != nil {
+		tr.lggr.Errorf("failed to handle txes by state: %v", err)
+		return
+	}
+	if tr.AbandonedTxCount() == 0 {
+		tr.lggr.Info("no abandoned txes found, skipping runLoop")
+		return
+	}
+	tr.lggr.Infof("%d abandoned txes found, starting runLoop", tr.AbandonedTxCount())
+
 	ttlExceeded := time.NewTicker(tr.ttl)
 	defer ttlExceeded.Stop()
 	for {
 		select {
 		case <-tr.mb.Notify():
 			for {
-				if tr.ctx.Err() != nil {
-					return
-				}
-				blockHeight, exists := tr.mb.Retrieve()
-				if !exists {
+				blockHeight := tr.mb.RetrieveLatestAndClear()
+				if blockHeight == 0 {
 					break
 				}
-				if err := tr.HandleTxesByState(tr.ctx, blockHeight); err != nil {
-					tr.lggr.Errorw(fmt.Errorf("failed to handle txes by state: %w", err).Error())
+				tr.lggr.Infof("received blockHeight %v", blockHeight)
+				if err := tr.handleTxesByState(ctx, blockHeight); err != nil {
+					tr.lggr.Errorf("failed to handle txes by state: %v", err)
+					return
+				}
+				if tr.AbandonedTxCount() == 0 {
+					tr.lggr.Info("all abandoned txes handled, stopping runLoop")
+					return
 				}
 			}
 		case <-ttlExceeded.C:
 			tr.lggr.Info("ttl exceeded")
-			tr.MarkAllTxesFatal(tr.ctx)
+			tr.markAllTxesFatal(ctx)
```

Comment:
Does this mean all txs have the same ttl regardless of how old/new they are?

Reply:
Yes correct